src/vm/checker/typecheck/natives/mod.rs

```diff
@@ -0,0 +1,320 @@
+use vm::errors::{ErrType as InterpError};
+use vm::functions::NativeFunctions;
+use vm::representations::{SymbolicExpression};
+use vm::types::{TypeSignature, AtomTypeIdentifier, TupleTypeSignature};
+use super::{TypeChecker, TypingContext, TypeResult, FunctionType, no_type, check_atomic_type}; 
+use vm::checker::errors::{CheckError, CheckErrors, CheckResult};
+
+mod maps;
+
+pub enum TypedNativeFunction {
+    Special(SpecialNativeFunction),
+    Simple(SimpleNativeFunction)
+}
+
+pub struct SpecialNativeFunction(&'static Fn(&mut TypeChecker, &[SymbolicExpression], &TypingContext) -> TypeResult);
+pub struct SimpleNativeFunction(pub FunctionType);
+
+fn arithmetic_type(variadic: bool) -> FunctionType {
+    if variadic {
+        FunctionType::Variadic(TypeSignature::new_atom( AtomTypeIdentifier::IntType ),
+                               TypeSignature::new_atom( AtomTypeIdentifier::IntType ))
+    } else {
+        FunctionType::Fixed(vec![TypeSignature::new_atom( AtomTypeIdentifier::IntType ),
+                                 TypeSignature::new_atom( AtomTypeIdentifier::IntType )],
+                            TypeSignature::new_atom( AtomTypeIdentifier::IntType ))
+    }
+}
+
+fn arithmetic_comparison() -> FunctionType {
+    FunctionType::Fixed(vec![TypeSignature::new_atom( AtomTypeIdentifier::IntType ),
+                             TypeSignature::new_atom( AtomTypeIdentifier::IntType )],
+                        TypeSignature::new_atom( AtomTypeIdentifier::BoolType ))    
+}
+
+fn check_special_list_cons(checker: &mut TypeChecker, args: &[SymbolicExpression], context: &TypingContext) -> TypeResult {
+    let typed_args = checker.type_check_all(args, context)?;
+    TypeSignature::parent_list_type(&typed_args)
+        .map_err(|x| {
+            let error_type = match x.err_type {
+                InterpError::BadTypeConstruction => CheckErrors::ListTypesMustMatch,
+                InterpError::ListTooLarge => CheckErrors::ConstructedListTooLarge,
+                InterpError::ListDimensionTooHigh => CheckErrors::ConstructedListTooLarge,
+                _ => CheckErrors::UnknownListConstructionFailure
+            };
+            CheckError::new(error_type)
+        })
+}
+
+fn check_special_print(checker: &mut TypeChecker, args: &[SymbolicExpression], context: &TypingContext) -> TypeResult {
+    if args.len() != 1 {
+        return Err(CheckError::new(CheckErrors::IncorrectArgumentCount(1, args.len())))        
+    }
+    
+    checker.type_check(&args[0], context)
```

Comment:
For `print`, do we want the first argument to be a string?  Or do all types have a string representation that can be used here?

Reply:
Moreover, for `print`, do we want to require that only printable ASCII characters get printed?  I.e. by transforming control characters into `^$X` for some `SX`?

Reply:
Right now, print accepts any value -- it simply uses the `fmt::Display` implementation on the Value enum, and I think that makes the most sense. The output formatting, though, is _pretty_ weird for strings (it prints the buffers).

Reply:
This makes it hard to print a mixture of debugging messages and values though, per my examples below.  What's the best way to do printf-style debugging in your opinion?

Reply:
Hmm, well, we could turn print into something like the rust `println!` function -- where we constrain the first argument to be an ASCII string, and then it takes a variable number of additional arguments of any type. The return type would have to be void then.